solver-next/scheduler_test.go

```diff
@@ -1422,7 +1422,378 @@ func TestMultipleCacheSources(t *testing.T) {
 	require.Equal(t, int64(0), cacheManager2.loadCounter)
 
 	require.NoError(t, j1.Discard())
+	j1 = nil
+}
+
+func TestRepeatBuildWithIgnoreCache(t *testing.T) {
+	t.Parallel()
+	ctx := context.TODO()
+
+	l := NewJobList(SolverOpt{
+		ResolveOpFunc: testOpResolver,
+	})
+	defer l.Close()
+
+	j0, err := l.NewJob("j0")
+	require.NoError(t, err)
+
+	defer func() {
+		if j0 != nil {
+			j0.Discard()
+		}
+	}()
+
+	g0 := Edge{
+		Vertex: vtx(vtxOpt{
+			name:         "v0",
+			cacheKeySeed: "seed0",
+			value:        "result0",
+			inputs: []Edge{
+				{Vertex: vtx(vtxOpt{
+					name:         "v1",
+					cacheKeySeed: "seed1",
+					value:        "result1",
+				})},
+			},
+		}),
+	}
+	g0.Vertex.(*vertex).setupCallCounters()
+
+	res, err := j0.Build(ctx, g0)
+	require.NoError(t, err)
+	require.Equal(t, unwrap(res), "result0")
+	require.Equal(t, int64(2), *g0.Vertex.(*vertex).cacheCallCount)
+	require.Equal(t, int64(2), *g0.Vertex.(*vertex).execCallCount)
+
+	require.NoError(t, j0.Discard())
+	j0 = nil
+
+	// rebuild with ignore-cache reevaluates everything
+
+	j1, err := l.NewJob("j1")
+	require.NoError(t, err)
+
+	defer func() {
+		if j1 != nil {
+			j1.Discard()
+		}
+	}()
+
+	g1 := Edge{
+		Vertex: vtx(vtxOpt{
+			name:         "v0",
+			cacheKeySeed: "seed0",
+			value:        "result0-1",
+			ignoreCache:  true,
+			inputs: []Edge{
+				{Vertex: vtx(vtxOpt{
+					name:         "v1",
+					cacheKeySeed: "seed1",
+					value:        "result1-1",
+					ignoreCache:  true,
+				})},
+			},
+		}),
+	}
+	g1.Vertex.(*vertex).setupCallCounters()
+
+	res, err = j1.Build(ctx, g1)
+	require.NoError(t, err)
+	require.Equal(t, unwrap(res), "result0-1")
+	require.Equal(t, int64(2), *g1.Vertex.(*vertex).cacheCallCount)
+	require.Equal(t, int64(2), *g1.Vertex.(*vertex).execCallCount)
+
+	require.NoError(t, j1.Discard())
+	j1 = nil
+
+	// ignore-cache in child reevaluates parent
+
+	j2, err := l.NewJob("j2")
+	require.NoError(t, err)
+
+	defer func() {
+		if j2 != nil {
+			j2.Discard()
+		}
+	}()
+
+	g2 := Edge{
+		Vertex: vtx(vtxOpt{
+			name:         "v0",
+			cacheKeySeed: "seed0",
+			value:        "result0-2",
+			inputs: []Edge{
+				{Vertex: vtx(vtxOpt{
+					name:         "v1",
+					cacheKeySeed: "seed1",
+					value:        "result1-2",
+					ignoreCache:  true,
+				})},
+			},
+		}),
+	}
+	g2.Vertex.(*vertex).setupCallCounters()
+
+	res, err = j2.Build(ctx, g2)
+	require.NoError(t, err)
+	require.Equal(t, unwrap(res), "result0-2")
+	require.Equal(t, int64(2), *g2.Vertex.(*vertex).cacheCallCount)
+	require.Equal(t, int64(2), *g2.Vertex.(*vertex).execCallCount)
+
+	require.NoError(t, j2.Discard())
+	j2 = nil
+}
```

Comment:
Do we need a test for `v0.ignoreCache=true, v1.ignoreCache=false`?

Reply:
You mean the non-parallel case for that? A parallel case for that should be covered in `TestParallelBuildsIgnoreCache`. I think it could be useful for making sure that cache is still stored if you do `ignorecache=true`.